modules/stochastic_tools/src/surrogates/PolynomialChaos.C

```diff
@@ -20,51 +21,93 @@ PolynomialChaos::validParams()
   InputParameters params = SurrogateModel::validParams();
   params.addClassDescription("Computes and evaluates polynomial chaos surrogate model.");
 
-  params.addRequiredParam<unsigned int>("order", "Maximum polynomial order.");
-  params.addRequiredParam<std::vector<DistributionName>>(
+  // Training parameters
+  params.addParam<SamplerName>("training_sampler", "Training set defined by a sampler object.");
+  params.addParam<VectorPostprocessorName>(
+      "results_vpp", "Vectorpostprocessor with results of samples created by trainer.");
+  params.addParam<std::string>(
+      "results_vector",
+      "Name of vector from vectorpostprocessor with results of samples created by trainer");
+  params.addParam<unsigned int>("order", "Maximum polynomial order.");
+  params.addParam<std::vector<DistributionName>>(
       "distributions", "Names of the distributions samples were taken from.");
 
+  params.addParamNamesToGroup("training_sampler results_vpp results_vector order distributions",
+                              "training");
   return params;
 }
 
 PolynomialChaos::PolynomialChaos(const InputParameters & parameters)
   : SurrogateModel(parameters),
-    _order(getParam<unsigned int>("order")),
-    _quad_sampler(nullptr),
-    _tuple(generateTuple(getParam<std::vector<DistributionName>>("distributions").size(), _order)),
-    _ncoeff(_tuple.size()),
-    _coeff(_ncoeff)
+    _tuple(declareModelData<std::vector<std::vector<unsigned int>>>("_tuple")),
+    _coeff(declareModelData<std::vector<Real>>("_coeff")),
+    _poly(declareModelData<std::vector<std::unique_ptr<const PolynomialQuadrature::Polynomial>>>(
+        "_poly"))
 {
-  MooseUtils::linearPartitionItems(
-      _ncoeff, n_processors(), processor_id(), _n_local_coeff, _st_local_coeff, _end_local_coeff);
 }
 
 void
 PolynomialChaos::initialSetup()
 {
-  SurrogateModel::initialSetup();
+  if (isTraining())
+  {
+    // Setup data needed for training
+    _order = getParam<unsigned int>("order");
```

Comment:
_order is not set if isTraining is false, but all the tests pass. How is this possible?

Reply:
Getting lucky? Turn on Valgrind testing and see if this fails with an uninitialized error.

Reply:
We figure it is initialized to something large. The way it is used would just result in a huge amount of data being unused. I will have a fix very soon.